### PR TITLE
Zero data loss disaster recovery using amazon rds

### PR DESCRIPTION
**Pull request by psammy-github**
# Content Pull Request Checklist

Please check to make sure your content submission meets the following requirements by adding an `x` between the brackets:

- [ ] The title of this pull request reflects the title of my content.
- [ ] I have reviewed the [Content Review Checklist](/content/blob/main/CONTENT_REVIEW_CHECKLIST.md).
- [ ] This pull request contains one piece of content (post or page).

**Important: if you are submitting a PR to update existing content, please ensure you [sync'ed your fork](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork) first before you created your branch.**

Is this pull request linked to a Content Proposal or Idea Suggestion Issue? If so, please link to it below using the number or full link.

```text
Resolves #5

or

Content for https://github.com/build-on-aws/content/issues/5
```

**By submitting this pull request, I confirm I own this content, have permission to use each image included in this content, and that it will be released under the [CC BY-SA 4.0 SA License](/LICENSE).**

### CODE REVIEW

**Comment by psammy-github** [comment]
ready for review

**Comment by setheliot** [comment]
This is content for the Resilience Wave

File: posts/zero-dataloss-dr-using-rds/index.md [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+| Amazon RDS for MySQL       | Log Sequence Number (LSN)  | Binary Log           |
+| Amazon RDS for PostgresSQL | Log Sequence Number (LSN)  | Transaction Log      |
+| Amazon RDS for MariaDB     | Log Sequence Number (LSN)  | Binary Log           |
+| Amazon RDS for Oracle      | System Change Number (SCN) | Redo Log             |
+| Amazon RDS for SQL Server  | Log Sequence Number (LSN)  | Transaction Log      |
+
+
+
+Cloud Enabled Databases capture all changes in logs also known as transaction, binary or redo logs depending on the database engine. These logs are generally write ahead logs (WAL) so all data manipulation is first persisted to these logs then the underlying database files using a process known as Check pointing. This provides a mechanism for a database to roll forward all SCN or LSN changes in the WAL(s) and recover its underlying files to be consistent after an ungraceful outage or server crash. Though any inflight transactions would be rolled back as they were not committed, this is a generally accepted model amongst Cloud Enabled database vendors and applications which results in Zero Data Loss.
```
Comment (setheliot) [inline posts/zero-dataloss-dr-using-rds/index.md:71]: This is great info.  Can you add a link or two inline for anyone who wants to learn more on these concepts?

File: posts/zero-dataloss-dr-using-rds/index.md [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+  
+![RDS crash recovery](images/pic5_recovery_rds.jpg "Figure 5. RDS crash recovery")
```
Comment (setheliot) [inline posts/zero-dataloss-dr-using-rds/index.md:73]: * There is no figure 4 (I think you need to renumber these)
* Please reference the figures, using the figure numbers, from the text. It is not clear to me which text refers to your figures 5 and 6

File: posts/zero-dataloss-dr-using-rds/index.md [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+  
+![Aurora crash recovery](images/pic6_recovery_aurora.jpg "Figure 6. Aurora crash recovery")
+  
+## High Availability vs Backups vs Disaster Recovery 
+One of the strengths of Amazon RDS is a simple and configurable Backup Solution, High Availability and Disaster Recovery options. This provides the capability to reinstate a failed database environment. But there are clear differences in their implementation, usage and potential for data loss.
+
+The industry standard term for the amount of data loss a system can tolerate is known as the Recovery Point Objective (RPO). This is measured as a unit of time it could be seconds, hours or even days. An RPO of 0 would mean that a database could not tolerate any data loss.
```
Comment (setheliot) [inline posts/zero-dataloss-dr-using-rds/index.md:82]: consider linking  Recovery Point Objective (RPO) to https://docs.aws.amazon.com/wellarchitected/latest/reliability-pillar/disaster-recovery-dr-objectives.html -- since you do not define it here

File: posts/zero-dataloss-dr-using-rds/index.md [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+High Availability is resilience to failure of individual services such as compute and storage. AWS by nature is highly available, if an EC2 fails due to an h/w issue this can be respawned on new h/w. EBS disks are resilient and striped. So we can expect Amazon RDS to be highly available as it's built using this same reliable, fault tolerant infrastructure. As we have discussed earlier if there is an issue that crashes the database, the database should still restart with zero data loss. If an RDS Database is setup to run as Multi AZ, RDS will automatically failover to a Disaster Recovery Site if one exists as well under many conditions. Using Multi AZ will provide an RPO of 0 due to synchronous replication.
```
Comment (setheliot) [inline posts/zero-dataloss-dr-using-rds/index.md:84]: optional

consider changing 

> AWS by nature is highly available

to

> AWS by nature is highly available ([as long as best practices are followed](https://docs.aws.amazon.com/whitepapers/latest/disaster-recovery-workloads-on-aws/shared-responsibility-model-for-resiliency.html))

Comment (setheliot) [reply posts/zero-dataloss-dr-using-rds/index.md:84]: > If an RDS Database is setup to run as Multi AZ, RDS will automatically failover to a Disaster Recovery Site if one exists 

I am not sure I understand this.

If it is setup to run as Multi-AZ then the recovery AZ will exist.  So what do you man by "if one exists"?

Also maybe specify that the disaster recovery site in this case is the AZ, just to make it clear

File: posts/zero-dataloss-dr-using-rds/index.md [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+
+Backups in Amazon RDS are based on EC2 Storage level Snapshots and for Cloud Enabled offerings this is coupled with transaction log backups every 5 minutes all automated when enabled and implicitly written to S3 Object bucket Storage. The initial Snapshot of a database will perform a full storage backup with subsequent Snapshots being forever incremental copying only changed disk blocks for speed and efficiency.
```
Comment (setheliot) [inline posts/zero-dataloss-dr-using-rds/index.md:86]: "EC2 instance storage level..." may be clearer

File: posts/zero-dataloss-dr-using-rds/index.md [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+Amazon RDS supports 5 different database engines as below:
+
+![Cloud Native vs Cloud Enabled](images/pic1_enabled_vs_native.jpg "Figure 1. Cloud Native vs Cloud Enabled")
+
+These Amazon RDS Databases engines generally fall into 2 categories
+- Cloud Enabled
+- Cloud Native
+
+Cloud Enabled Databases run using similar infrastructure and setup as if they were running on premise. In a way they are databases designed for an on premise environment and tailored to run on the cloud and utilise those great services that AWS provides. Amazon RDS which caters for engines Oracle, SQL Server, Postgres, MySQL and MariaDB can be considered Cloud Enabled offerings.
```
[the remark this answers is not in the document]
Comment (setheliot) [reply posts/zero-dataloss-dr-using-rds/index.md:45]: ok, I see you use the cloud enabled and cloud native throughout the remainder of the article.  I do not want to cause a major disruption.  If you are happy with it, and cannot think of a replacement, then I am fine with it

File: posts/zero-dataloss-dr-using-rds/index.md [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+  
+| Backup                              | Cadence                 |
+|-------------------------------------|-------------------------|
+| Full Physical                       | One off Initial Backup  |
+| Incremental Physical                | Daily Backup Window     |
+| Cloud Enabled Transactional Logical | Every 5 Minutes         |
+| Cloud Native Transactional Logical  | Continuous              |
+
+  
+The Hybrid Cloud Native Amazon Aurora backups are continuous and based on its intelligent storage clustered setup and how data changes are recorded, due to this we should be able to restore an Aurora database with no data loss as there is no transaction log backup needed, the transactions are part of the data persisted to disk across all 3 AZs mentioned earlier.
```
Comment (setheliot) [inline posts/zero-dataloss-dr-using-rds/index.md:96]: would "persisted to disks across all 3 AZs mentioned earlier" (plural disks) be clearer?

**Review by setheliot — CHANGES_REQUESTED** [review]

File: posts/zero-dataloss-dr-using-rds/index.md [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+- Synchronous using Oracle RDS Custom, where we are able to access the underlying operating system and also make certain customisations that are not possible under normal Amazon RDS Regular. One of the customisations that is possible for Oracle is to setup for logical synchronous replication.
+ 
+## Database support for Logical Synchronous Replication
+We can see below which database engines support SYNC and SEMISYNC replication so therefore potentially an RPO 0 depending on the replication type.
+
+
+| Database Engine              | Storage Sub System | Multi AZ SAN Physical Replication Support (SYNC) | Multi AZ DB Replica Logical Replication Support (SYNC) | Multi AZ DB Replica Logical Replication Support (SEMI-SYNC)  |
+|------------------------------|--------------------|--------------------------------------------------|--------------------------------------------------------|--------------------------------------------------------------|
+| Amazon Aurora MySQL          | Multi AZ Clustered | Implicit                                         | Yes                                                    | No                                                           |
```
Comment (setheliot) [inline posts/zero-dataloss-dr-using-rds/index.md:150]: are you sure that Aurora Multi-AZ replication is synchronous?   This indicates it is asynchronous  https://aws.amazon.com/rds/features/read-replicas/

**Review by setheliot — CHANGES_REQUESTED** [review]

**Review by setheliot — APPROVED** [review]

File: posts/zero-dataloss-dr-using-rds/index.md [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+  - https://aws.amazon.com/rds/features/multi-az/
+  - https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Concepts.MultiAZ.html     
+  - https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/Concepts.MultiAZSingleStandby.html     
+  - https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/multi-az-db-clusters-concepts.html     
+  - https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_SQLServerMultiAZ.html     
+  - https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/custom-sqlserver-multiaz.html     
+  - https://aws.amazon.com/what-is/disaster-recovery/     
+  - https://docs.aws.amazon.com/wellarchitected/latest/reliability-pillar/welcome.html
+- White Papers
+  - https://docs.aws.amazon.com/whitepapers/latest/disaster-recovery-workloads-on-aws/disaster-recovery-options-in-the-cloud.htm
```
[the remark this answers is not in the document]
Comment (psammy-github) [reply posts/zero-dataloss-dr-using-rds/index.md]: Fixed and all links are not MD compliant

File: posts/zero-dataloss-dr-using-rds/index.md [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+
+## What does Zero Data Loss Disaster Recovery (ZDLDR) actually mean and why would it be needed?
+Disaster Recovery is the act of recovering an impaired service after a failure event that effects the primary Availability Zone(s), Data Centre(s) or Region(s). That event could range from natural disasters, power cuts, and network outages to political activism stopping or impacting the data centres where servers and services are physically run from.
+
+Recovery from failure and Testing of recovery procedures also forms part of the [AWS Well-Architected Framework](https://docs.aws.amazon.com/wellarchitected/latest/reliability-pillar/plan-for-disaster-recovery-dr.html) falling under the 'Reliability' Pillar. This is a core component of any AWS architected design so should be well understood.
+
+ > Your resiliency strategy should also include Disaster Recovery (DR) objectives based on strategies to recover your workload in case of a disaster event.
+
+ Amazon RDS Disaster Recovery Data Loss scope is at the database level and measureable in time or database system change. We recovery the database and everything contained within to the same consistent point.
```
[the remark this answers is not in the document]
Comment (psammy-github) [reply posts/zero-dataloss-dr-using-rds/index.md]: - It should be recover, updated
- I've added a sentence to explain what database system change is and it will be discussed later in blog

File: posts/zero-dataloss-dr-using-rds/index.md [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+  
+![RDS Multi AZ EBS](images/pic2_az_ebs.jpg "Figure 2. RDS Multi AZ EBS")
+  
+
+Cloud Native Databases are designed from the ground up to utilise cloud features that are generally only available on the cloud. Amazon Aurora Postgres and Amazon Aurora MySQL (both of which are still part of the RDS family) can be considered a hybrid cloud native database offering. Aurora decouples the compute from the storage layer. The storage layer has extra efficiencies, safe guards and utilises multiple AZs to enhance high availability. 
+
+As depicted below Amazon Aurora RDS writes to a clustered storage volume that spans across all 3 Availability Zones where 6 copies of the data will be present. The data is striped such that an Aurora database can be stood up for Disaster Recovery at any time in any Availability Zone within the same Region with no further replication or overhead. 
+  
+![RDS Multi AZ Clustered](images/pic2_az_clustered.jpg "Figure 3. RDS Multi AZ Clustered")
```
[the remark this answers is not in the document]
Comment (psammy-github) [reply posts/zero-dataloss-dr-using-rds/index.md:56]: Very good point, this has been updated in text and diagram, diagram will be show when I perform a new pull, for some  reason its not showing on the automated pull preview

File: posts/zero-dataloss-dr-using-rds/index.md (hunk 1 above)
Comment (psammy-github) [reply posts/zero-dataloss-dr-using-rds/index.md]: added a link to wikipedia on WAL

File: posts/zero-dataloss-dr-using-rds/index.md (hunk 2 above)
Comment (psammy-github) [reply posts/zero-dataloss-dr-using-rds/index.md]: images all relabelled to have an correct order with no gaps

File: posts/zero-dataloss-dr-using-rds/index.md (hunk 3 above)
Comment (psammy-github) [reply posts/zero-dataloss-dr-using-rds/index.md]: Done

File: posts/zero-dataloss-dr-using-rds/index.md (hunk 4 above)
Comment (psammy-github) [reply posts/zero-dataloss-dr-using-rds/index.md]: - Added link for HA best practice and updated wording
- Agree the Multi AZ sentence is worded badly I've changed it

File: posts/zero-dataloss-dr-using-rds/index.md (hunk 7 above)
Comment (psammy-github) [reply posts/zero-dataloss-dr-using-rds/index.md]: I've Updated text for same

File: posts/zero-dataloss-dr-using-rds/index.md (hunk 8 above)
Comment (psammy-github) [reply posts/zero-dataloss-dr-using-rds/index.md]: This came as a surprise to me, text and tables updated to reflect the same

**Review by TimGustafson — CHANGES_REQUESTED** [review]
I will send an e-mail with more details.

File: posts/zero-dataloss-dr-using-rds/index.md (hunk 5 above)
Comment (psammy-github) [reply posts/zero-dataloss-dr-using-rds/index.md]: updated

File: posts/zero-dataloss-dr-using-rds/index.md [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+
+To restore an RDS Cloud enabled database would need a restore of the full backup + incremental backup(s) + for cloud enabled database the roll forward of the transaction logs. Potentially the RPO using Amazon cloud enabled RDS backups would be 5 minutes at most (the cadence of transaction log backups). This would only be applicable if not using Multi AZ to reinstate the database.
+
+To restore an RDS hybrid cloud native database would need a restore of the full backup + incremental backup(s), there are no transaction logs to restore, we should expect an RPO of 0 under normal circumstances as the data is highly available and replicated, though as noted above backups could be 5 minutes behind, so as long as Aurora is still available that 5 minutes is less of an issue due to the high availability of data across 3 availability zones.
```
Comment (damonk10) [inline posts/zero-dataloss-dr-using-rds/index.md:101]: This needs to be broken into at least 2 sentences (probably 3). It's just very hard to follow.

File: posts/zero-dataloss-dr-using-rds/index.md [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+
+## RDS Replicas
+
+A replica in Amazon RDS or Amazon Aurora represents a database that also has the potential to be used as an additional read only copy of the live database while still being continually updated with all changes. Using Amazon RDS Oracle, there is an extra license fee to provision a read-only replica known as Active Dataguard. Hence with this database engine we can also have the replica in recovery mode only not accessible for read only access where that license fee is then not required this is called a mounted replica. Amazon Aurora replicas are based on the decoupled clustered storage that is automatically replicated across all AZs, but the replication thereafter is considered asynchronous as the replicas are kept updated with native database replication. For disaster recovery, though, Aurora would rely on its decoupled storage for an RPO of 0. Furthermore, for Amazon Aurora we don't even need to stand up a replica as the potential to create one is still there at any time as the data is clustered across all available AZs to facilitate that.
```
Comment (damonk10) [inline posts/zero-dataloss-dr-using-rds/index.md:138]: Not tracking with this sentence: "Hence with this database engine we can also have the replica in recovery mode only not accessible for read only access where that license fee is then not required this is called a mounted replica." Probably break it into two simpler sentences.

File: posts/zero-dataloss-dr-using-rds/index.md [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+  
+![RDS crash recovery](images/pic5_recovery_rds.jpg "Figure 4. RDS crash recovery")
+  
+RDS Cloud Native databases, namely Amazon Aurora, use a clustered log structured distributed storage system. This means once a change has occurred in the database, the log records are sent to the storage tier and added to an in-memory queue. The database is then freed from any other overhead regarding management of these records. It falls to the Storage tier to take responsibility for persisting those records to disk and data page updates to reflect the change. As there is no transaction log or WAL as such on the database tier, the crash recovery phase is minimal when recovering from an ungraceful shutdown or crash. The data on disk is always up-to-date or log records to recreate current images exist on disk to be applied in parallel and asynchronously this results in Zero Data Loss. This is depicted below in Figure 5. It should also be noted Auroras database working memory area, called a page cache, is also decoupled from the database so it can independently survive a database crash and be reused once the database is back.
```
[the remark this answers is not in the document]
Comment (setheliot) [reply posts/zero-dataloss-dr-using-rds/index.md]: This sentence was removed in a subsequent edit

File: posts/zero-dataloss-dr-using-rds/index.md (hunk 12 above)
Comment (setheliot) [reply posts/zero-dataloss-dr-using-rds/index.md]: done

File: posts/zero-dataloss-dr-using-rds/index.md (hunk 13 above)
Comment (setheliot) [reply posts/zero-dataloss-dr-using-rds/index.md]: fixed

**Comment by lenadroid** [comment]
Seems like PR is approved and can be merged @cherbk @setheliot
